#42 battery characteristics

adiojha629 opened the pull request:
Closes #42
Closes #37

Goal: Have the ThrustModel function only take inputs of SimData.
We need to augment SimData to record information on how much capacity is left in the battery
Variables Defined:

SimData.battery_ratio %ratio between current battery capacity and full capacity (dimensionless)
SimData.battery_update %change in battery capacity in mAh

In order to pass the value of battery_update to SimData, the output of Derivatives (kOut) has a term called battery update

ie kOut.battery_update = ThrustModel(SimData)(3)
This kOut.battery_update is used in RK4 integration to update SimData.battery_update.

The Thrust model outputs 3 values:

Thrust_x
Thrust_y
battery_update (in mAh)

adiojha629 committed:
Thrust uses SimData structure

adiojha629 committed:
Created get_Prop_AoA

In order to find torque we need to know the propellor's angle of attack. I created a function to do just that.

I added a structure to SimData called SimData.Prop. It holds variables that we need for the Propellor

adiojha629 committed:
Using P=Fv to get current draw

adiojha629 committed:
Ishan stuff

adiojha629 committed:
ThurstModel with only SimData input

Goal: Have the ThrustModel function only take inputs of SimData.
We need to augment SimData to record information on how much capacity is left in the battery
Variables Defined:

SimData.battery_ratio %ratio between current battery capacity and full capacity (dimensionless)
SimData.battery_update %change in battery capacity in mAh

In order to pass the value of battery_update to SimData, the output of Derivatives (kOut) has a term called battery update

ie kOut.battery_update = ThrustModel(SimData)(3)
This kOut.battery_update is used in RK4 integration to update SimData.battery_update.

The Thrust model outputs 3 values:

Thrust_x
Thrust_y
battery_update (in mAh)

TheGreatFez commented on Simulation/RK4_Integration.m:
```diff
@@ -5,10 +5,10 @@
     % Initialize with zeroed delta
     k0.P_dot = [0,0]; % Position Derivative
-    k0.V_dot = [0,0]; % Velocity Derivative
+    k0.V_dot = [0,0]; % Velocity Derivative thrust here 
```
I don't think this was meant to be kept here? Or at least it doesn't read properly

TheGreatFez commented on Simulation/RK4_Integration.m:
```diff
@@ -5,10 +5,10 @@
     k0.O_dot = [0];   % Orientation Derivative
     k0.Time    = SimData.Time;
-
+    %x = ThrustModel(SimData)
```
Remove commented out code

TheGreatFez commented on Simulation/TSoG_X1_Sim.m:
```diff
@@ -80,8 +88,7 @@
     SimData = RK4_Integration(SimData);
-
-
+    %x = ThrustModel(SimData)  
```
Remove commented out code

TheGreatFez commented on Simulation/ThrustModel.m:
```diff
@@ -18,14 +18,32 @@
+  RPM = Kv*voltage;%RPM based on Kv and Voltage
+  
+  Thrust = 4.392399*10^(-8)*RPM*diameter^(3.5)*pitch^(-.5)*((4.23333*10^(-4))*RPM*pitch-airspeed);%calculate thrust (based on model in excel file)
+  #Thrust= 10*Throttle*(1-airspeed/50)
```
Remove commented out code

TheGreatFez commented on Simulation/TSoG_X1_Sim.m:
```diff
@@ -92,8 +99,12 @@
+    
+    #updates on battery for the next iteration
+    Results.battery_status(i) = Results.battery_status(i-1) - SimData.battery_update;
+    SimData.battery_ratio = Results.battery_status(i) / SimData.battery_total_cap;
```
A LiPo battery actually goes from full charge (12.6V) to min charge (~9V). 

If we use this "battery_ratio" the thrust will reduce linearly until the battery capacity is 0. However, LiPo batteries can output up to 11.1V all the way down to 15% capacity ([Source](https://blog.ampow.com/lipo-voltage-chart/))

I have made a comment as well that we need to adjust this since after doing some further research the Max Volatage is actually 12.6V not 11.1V as previously thought.

Nufflee replied:
A thing to keep in mind is that these numbers are only valid for 3S LiPos. The usual formula is `S * 3.7V` where S is the number of batteries in series and 3.7V is the nominal voltage (average of max and min voltage) for a regular LiPo.
And yeah, the discharge curves of LiPos are not linear ([example](https://electronics.stackexchange.com/questions/32321/lipoly-battery-when-to-stop-draining)).

TheGreatFez commented on Simulation/ThrustModel.m:
```diff
@@ -18,14 +18,32 @@
+  Throttle = battery_ratio * Raw_throttle; % a percent representing the power available in the battery
+  
+  
+  voltage = Max_Volt * Throttle; %The voltage on the motor (a scaled down version of the max voltage)
```
After some further research, I realized/remembered that Max Voltage of a 3S LiPo battery is actually 12.6V not 11.1V (See other comment regarding this)

TheGreatFez reviewed (changes requested):
Most comments are for removal of commented out code or clean up

However, the battery_update logic needs some update to how LiPo batteries work. Shouldn't be too bad though!

Nufflee commented on Simulation/Derivatives.m:
```diff
@@ -17,8 +17,13 @@
+  battery_update = Thrust(3);
+  
+  Thrust = [Thrust(1),Thrust(2)];
+  F = Drag + Lift + Thrust; % Net force on the object Commented out b/c Thrust causes imaginary numbers
```
I think you meant to remove `Commented out b/c Thrust causes imaginary numbers
`

Nufflee commented on Simulation/TSoG_X1_Sim.m:
```diff
@@ -35,7 +35,15 @@
+  
+  
+  %Battery Variables
+  SimData.battery_status = 1300;       %Current Status of the battery is (mAh)
```
`battery_status` implies the status/state of charge. A better name here would be something like `battery_capacity`.

Nufflee replied:
Battery related variables should also probably be a part of the `Plane` substructure.

Nufflee commented on Simulation/ThrustModel.m:
```diff
@@ -18,14 +18,32 @@
-  RPM = Max_Volt*Kv*Throttle;%use Throttle to scale down the max RPM 
-  Thrust = 4.392399*10^(-8)*RPM*diameter^(3.5)*pitch^(-.5)*((4.23333*10^(-4))*RPM*pitch-airspeed)%calculate thrust (based on model in excel file)
+  
+  %%How to get Throttle:
```
This comment seem a bit unnecessary

Nufflee commented on Simulation/TSoG_X1_Sim.m:
```diff
@@ -92,8 +99,12 @@
-
+    Results.ThrottleInput(i) = SimData.battery_ratio*SimData.TestCase.GetThrottle(SimData.Time);
+    
+    #updates on battery for the next iteration
```
Please stick to using `%` for comments everywhere for consistency.

Nufflee commented on Simulation/TSoG_X1_Sim.m:
```diff
@@ -92,8 +99,12 @@
+    Results.ThrottleInput(i) = SimData.battery_ratio*SimData.TestCase.GetThrottle(SimData.Time);
+    
+    #updates on battery for the next iteration
+    Results.battery_status(i) = Results.battery_status(i-1) - SimData.battery_update;
```
We should probably be updating the `battery_status/charge` somewhere in derivatives and remove `battery_update` from `SimData` because it really is the derivative of the battery charge.

TheGreatFez committed:
Merging with Master for latest changes

TheGreatFez committed:
Change from a 0-100% voltage to a voltage based on LiPo battery

Commit:
Updated ThrustModel output to 2 outputs. Updated battery_update to B_dot

Commit:
Removed comment from line 8. Added B_dot to kN structures. Added B_delta to avg_delta structure

Commit:
Updates to Thrust Model:
- Changed to 2 outputs (Thrust and B_dot)
- Changed to mAh/s on current draw
- Updated Voltage calculation to just use an average
- Changed Minimum Voltage to 3.2 for more safe battery usage

Commit:
Forgot to add /3600 to B_dot calculation

Nufflee commented on Simulation/ThrustModel.m:
```diff
@@ -5,27 +5,45 @@
+  Throttle = SimData.TestCase.GetThrottle(SimData.Time);
+  diff_Volt = Max_Volt - Min_Volt
+  voltage = Avg_Volt*Throttle; % Voltage sent to the motor
+  airspeed = norm(SimData.StateVector.Velocity)#linear velocity of plane in x,y plane
```
I think this comment is a bit redundant too. At least make it a `%` comment for consistency.

Nufflee commented on Simulation/ThrustModel.m:
```diff
@@ -5,27 +5,45 @@
+  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+  %%How to get Throttle:
+  battery_cap_percent = SimData.Plane.Cap/SimData.Plane.MaxCap;
```
`SimData.Cap` should probably be a more descriptive name like `SimData.BatteryCap` since "cap" could mean a lot of things.

Commit:
Convert power to a linear model based on throttle and max power output assuming capped by the Speed Controller

TheGreatFez committed:
Added battery cap update to Derivatives, set a min of 0 for capacity

TheGreatFez committed:
Added Battery Cap update to RK4_Integration, set min to 0 for capacity

TheGreatFez committed:
Added logic for 0 capacity (no power). Removed unneeded comment. Removed unneded calculations for voltage and battery cap percentage

TheGreatFez committed:
Moved Battery variables to Plane structure. Updated Results with Battery and MaxBattery

TheGreatFez committed:
Minor fixes to TempTestCase: Velocity starts positive now, Input Plots have correct axis limits. Added battery variables to Plane structure

TheGreatFez reviewed (approved):
Temp Approval to close out my review

Nufflee commented on Simulation/ThrustModel.m:
```diff
@@ -5,27 +5,48 @@
+    Thrust = 4.392399*10^(-8)*RPM*diameter^(3.5)*pitch^(-.5)*((4.23333*10^(-4))*RPM*pitch-airspeed);%calculate thrust (based on model in excel file)
+  endif
+    
+  #Calculations to update battery
```
`%` for comments.

TheGreatFez committed:
Change to % for comments. Removed one unneeded commented out code

Commit:
Removed unneeded comments
